core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -87,36 +85,11 @@
  * according to a dynamic programming algorithm.
  */
 public class VolcanoPlanner extends AbstractRelOptPlanner {
-  protected static final double COST_IMPROVEMENT = .5;
 
   //~ Instance fields --------------------------------------------------------
 
   protected RelSubset root;
 
-  /**
-   * If true, the planner keeps applying rules as long as they continue to
-   * reduce the cost. If false, the planner terminates as soon as it has found
-   * any implementation, no matter how expensive.
-   */
-  protected boolean ambitious = true;
-
-  /**
-   * If true, and if {@link #ambitious} is true, the planner waits a finite
-   * number of iterations for the cost to improve.
-   *
-   * <p>The number of iterations K is equal to the number of iterations
-   * required to get the first finite plan. After the first finite plan, it
-   * continues to fire rules to try to improve it. The planner sets a target
-   * cost of the current best cost multiplied by {@link #COST_IMPROVEMENT}. If
-   * it does not meet that cost target within K steps, it quits, and uses the
-   * current best plan. If it meets the cost, it sets a new, lower target, and
-   * has another K iterations to meet it. And so forth.
-   *
-   * <p>If false, the planner continues to fire rules until the rule queue is
-   * empty.
-   */
-  protected boolean impatient = false;
```

Comment:
Is there any way now to stop the planner search without search space exhaustive exploration? `impatient` flag was pretty useful for it.

Reply:
Nope. `impatient ` doesn't guarantee to find the best plan, no one ever used it.

Reply:
You are right, 'impatient' flag doesn't help to find the best plan. But it helps to interrupt the search if it takes too much time and you  are ok to go ahead with a sub optimal plan. We use it for this purpose.

Reply:
I think you should investigate why your planner takes too much time to generate a plan.

Reply:
We've investigated it and looks like it is very similar to the [abstract converters problem](https://issues.apache.org/jira/browse/CALCITE-2970). `impatient` flag was quite helpful for us in some cases.

Reply:
Can we rename `impatient` flag to [`useSimulatedAnnealing`](https://en.wikipedia.org/wiki/Simulated_annealing) ?
It would sound sophisticated enough to keep it from removing :)

Jokes aside, I think the planning cannot be always fit into a finite amount of time, so it makes sense to have a predefined timeout for the planner. That is it could give up and just return the best it has so far.

Reply:
It is easy to fix. Don't follow Calcite's way, it is doing wrong. 
- Transformation rule matches only logical nodes, not physical nodes.
- Create your rule to match AbstractConverter to convert to your own physical sort/distribute nodes.
- Customize your own convention to use abstract converter only when both from and to are physical convention.

If your system is not a federated system, limit your convention to 2: NONE and Physical one.

But since you already modified the code, you can still add it back in your own codebase even I removed this.

Reply:
Yes, there are many ways to fix this problem. But anyway, I  agree with Vladimir that we need an ability to restrict the planning time. It is a good point.

Reply:
Given that the optimizer is not exploring all the join order alternatives (the catalan number), if the planning can't fit into a finite amount of time, there is a bug.

Reply:
There is always a trade-off between planning time and plan quality. If I have a very complex query I don't want have the best plan ever with a very long planning  time. I want some 'acceptable' plan obtained in 'acceptable' time. 
It is a good practice to have an optimization timeout, especially in production systems. See optimizer  timeout for MS SQL Server.
`impatient` flag was the kind of this timeout. I'm not insist that we should keep exactly this flag, but throwing it away without an alternative is not a good idea, IMHO.

Reply:
I have added a timeout exception and return the current best plan immediately after timeout. This gives user more ability to customize when to timeout.

Reply:
As a user, you can extend VolcanoPlanner and override the checkCancel method.

Reply:
Ok, it might help us.
Off-topic question: why do we need several `VolcanoPlannerPhase` when only the one of them is used? May be we can throw it away along with the queue importance?

Reply:
In default implementation, there is only 1 phase used. But actually it might help to have multiple phases to split different kinds of rules.

Reply:
We should still keep planning phase. In classic volcano model, there are transformation, implementation and optimization phases. The planner can do specific things for different phases. It's not used now, but we could probably extend it in the future.